plugins/uast-kotlin-fir/tests/org/jetbrains/uast/test/env/kotlin/AbstractFirUastTest.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.test.env.kotlin
+
+import com.intellij.core.CoreApplicationEnvironment
+import com.intellij.openapi.extensions.Extensions
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.testFramework.LightProjectDescriptor
+import com.intellij.testFramework.registerServiceInstance
+import com.intellij.util.io.URLUtil
+import org.intellij.plugins.intelliLang.inject.InjectLanguageAction
+import org.jetbrains.kotlin.idea.test.KotlinLightCodeInsightFixtureTestCase
+import org.jetbrains.kotlin.idea.test.KotlinWithJdkAndRuntimeLightProjectDescriptor
+import org.jetbrains.kotlin.resolve.jvm.extensions.AnalysisHandlerExtension
+import org.jetbrains.uast.UFile
+import org.jetbrains.uast.UastFacade
+import org.jetbrains.uast.UastLanguagePlugin
+import org.jetbrains.uast.common.kotlin.FirUastPluginSelection
+import org.jetbrains.uast.kotlin.FirKotlinUastService
+import org.jetbrains.uast.kotlin.KotlinUastResolveProviderService
+import org.jetbrains.uast.kotlin.firKotlinUastPlugin
+import org.jetbrains.uast.kotlin.internal.CliKotlinUastResolveProviderService
+import org.jetbrains.uast.kotlin.internal.FirCliKotlinUastService
+import org.jetbrains.uast.kotlin.internal.UastAnalysisHandlerExtension
+import org.jetbrains.uast.kotlin.kotlinUastPlugin
+import java.io.File
+
+abstract class AbstractFirUastTest : KotlinLightCodeInsightFixtureTestCase(), FirUastPluginSelection {
+    private fun registerExtensionPointAndServiceIfNeeded() {
+        if (!isFirPlugin) return
+
+        val area = Extensions.getRootArea()
+        CoreApplicationEnvironment.registerExtensionPoint(
+            area,
+            UastLanguagePlugin.extensionPointName,
+            UastLanguagePlugin::class.java
+        )
+        area.getExtensionPoint(UastLanguagePlugin.extensionPointName)
+            .registerExtension(firKotlinUastPlugin, testRootDisposable)
```

Comment:
It'd be nice to introduce, e.g., `// FIR_IDENTICAL` (same as FIR diagnostics tests) when rendered text / logs become identical so that we can ignore files with same contents.

The bottlenecks are here: how to unregister extension points and service? If it's already there, let me know. If it's not there yet but doable, then we can do that in a separate PR/commit. If it's not that easy to do so, well, we need the current structure: running tests with two different sets of extension points and services.

Reply:
For unregistering extension point, see ExtensionPoint.unregisterExtension

Reply:
Aha, good to know that. Actually, after sending this, I realized that, with or without unregister, such directive and comparisons are still possible. I wonder, though, it's far from the current state. :)